File: source/downloads/code/LeetCode/Python/Minimum-Size-Subarray-Sum.py
class Solution(object):
    def minSubArrayLen(self, s, nums):
        """
        :type s: int
        :type nums: List[int]
        :rtype: int
        """
        start = 0
        end = 0
        size = len(nums)
        result = size+1
        sum = 0
        while end<size:
            # find the end
            while end<size and sum < s:
                sum += nums[end]
                end += 1
            # find the start
            while start<end and sum >= s:
                result = min(result,end-start)
                sum -= nums[start]
                start += 1
        if result <= size:
            return result
        else:
            return 0

File: downloads/code/LeetCode/Java/Minimum-Size-Subarray-Sum.java
public class Solution {
    public int minSubArrayLen(int s, int[] nums) {
        int sum = 0;
        int start = 0;
        int end = 0;
        int ret = Integer.MAX_VALUE;
        while(end<nums.length){
            // 找到一个序列，使sum>=s
            while(end<nums.length && sum < s){
                sum += nums[end];
                end++;
            }
            // 没有解了
            if(sum < s) break;

            // 寻找该序列的start
            while(start<end && sum>=s){
                sum -= nums[start];
                start++;
            }

            ret = Math.min(ret,end-start+1);
        }
        
        return ret==Integer.MAX_VALUE?0:ret;
    }
}
